File: 2024-08-16/new_package/mock.go
package main

import (
	"fmt"
	"sync"
	"time"
)

var m = sync.Mutex{}
var wg = sync.WaitGroup{}
var dbData = []string{"id1", "id2", "id3", "id4", "id5", "id6"}
var dbDataMap = map[string]string{
	"id1": "RandomString1",
	"id2": "AnotherRandomString",
	"id3": "YetAnotherString",
	"id4": "FourthRandomString",
	"id5": "FifthRandomString",
	"id6": "LastRandomString",
}

func main() {
	results := make(chan string)
	// done := make(chan bool)
	expectedResults := len(dbData)
	// moreResults := []string{}
	myCounter := 1

	t0 := time.Now()
	for _, id := range dbData {
		wg.Add(1)
		go func(id string) {
			results <- dbCall(id)
			// Without this lock, this block has inconsistent behavior
			// m.Lock()
			var thing = myCounter
			time.Sleep(100_000_000)
			myCounter = thing + 1
			// m.Unlock()
			wg.Done()
		}(id)
	}

	// go func() {
	// 	wg.Wait()
	// 	close(results)
	// }()

	i := 0
	for result := range results {
        fmt.Println("Received result:", result)
		i++
		if i == expectedResults {
			close(results)
		}
    }

	wg.Wait()
    // close(results)

	fmt.Printf("\nCounter results: %v", myCounter)
	fmt.Printf("\nTotal exeuction time: %v", time.Since(t0))
}

func dbCall(id string) string {
	// var delay float32 = rand.Float32() * 2000
	var delay = 2000
	time.Sleep(time.Duration(delay) * time.Millisecond)
	fmt.Println("The result from the database is:", id)
	return dbDataMap[id]
}

File: 2024-08-16/new_package/mock.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, Instant};

fn main() {
    let db_data = vec!["id1", "id2", "id3", "id4", "id5", "id6"];
    let mut db_data_map = HashMap::new();
    db_data_map.insert("id1", "RandomString1");
    db_data_map.insert("id2", "AnotherRandomString");
    db_data_map.insert("id3", "YetAnotherString");
    db_data_map.insert("id4", "FourthRandomString");
    db_data_map.insert("id5", "FifthRandomString");
    db_data_map.insert("id6", "LastRandomString");

    let db_data_map = Arc::new(db_data_map);
    let expected_results = db_data.len();
    let my_counter = Arc::new(Mutex::new(1));

    let (tx, rx) = std::sync::mpsc::channel();

    let t0 = Instant::now();

    let handles: Vec<_> = db_data
        .into_iter()
        .map(|id| {
            let tx = tx.clone();
            let db_data_map = Arc::clone(&db_data_map);
            let my_counter = Arc::clone(&my_counter);

            thread::spawn(move || {
                tx.send(db_call(&id, &db_data_map)).unwrap();
                
                // Without this lock, this block has inconsistent behavior
                // let mut counter = my_counter.lock().unwrap();
                let thing = *my_counter.lock().unwrap();
                thread::sleep(Duration::from_nanos(100_000_000));
                *my_counter.lock().unwrap() = thing + 1;
                // drop(counter);
            })
        })
        .collect();

    let mut i = 0;
    for result in rx.iter().take(expected_results) {
        println!("Received result: {}", result);
        i += 1;
        if i == expected_results {
            break;
        }
    }

    for handle in handles {
        handle.join().unwrap();
    }

    println!("\nCounter results: {}", *my_counter.lock().unwrap());
    println!("\nTotal execution time: {:?}", t0.elapsed());
}

fn db_call(id: &str, db_data_map: &HashMap<&str, &str>) -> String {
    let delay = Duration::from_millis(2000);
    thread::sleep(delay);
    println!("The result from the database is: {}", id);
    db_data_map[id].to_string()
}